File: python-algorithm/leetcode/problem_429.py
"""429. N-ary Tree Level Order Traversal
https://leetcode.com/problems/n-ary-tree-level-order-traversal/
"""
import collections
from typing import List


class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children


class Solution:
    def level_order(self, root: 'Node') -> List[List[int]]:
        if not root:
            return []
        ans = []
        q = collections.deque()
        q.append(root)
        while q:
            cur = []
            n = len(q)
            for i in range(n):
                node = q.popleft()
                cur.append(node.val)
                if node.children:
                    for child in node.children:
                        q.append(child)
            ans.append(cur)
        return ans

File: go-algo/leetcode/problem429.go
package main

// 429. N-ary Tree Level Order Traversal
// https://leetcode.com/problems/n-ary-tree-level-order-traversal/

func levelOrder(root *Node429) [][]int {
	ans := make([][]int, 0)
	if root == nil {
		return ans
	}
	q := []*Node429{root}
	for len(q) > 0 {
		n := len(q)
		level := make([]int, 0)
		for i := 0; i < n; i++ {
			node := q[0]
			level = append(level, node.Val)
			q = q[1:]
			for _, child := range node.Children {
				q = append(q, child)
			}
		}
		ans = append(ans, level)
	}
	return ans
}

type Node429 struct {
	Val      int
	Children []*Node429
}
